File: quant1x/contrib/data/tdx/level1/std/heartbeat.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from ...command import Command
from ... import protocol


class Heartbeat(protocol.BaseMessage):
    """心跳（合并Request和Response）"""
    def __init__(self):
        super().__init__(Command.STD_HEARTBEAT)
        self.info = ""
        self.request_header.packet_type = 0x02

    def serialize_request_body(self) -> bytes:
        return b""

    def deserialize_response_body(self, data: bytes) -> None:
        if len(data) >= 10:
            s = data[:10]
        else:
            s = data
        try:
            self.info = s.decode('gbk', errors='ignore').split('\x00', 1)[0]
        except Exception:
            self.info = s.decode('utf-8', errors='ignore').split('\x00', 1)[0]

File: quant1x/contrib/data/tdx/level1/std/heartbeat.rs
// -*- coding: utf-8 -*-
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// heartbeat — 心跳维持 (STD_HEARTBEAT, 0x0004)

use crate::std::BinaryStream;

use super::super::super::command::*;
use super::super::super::helpers::get_sequence_id;
use super::super::super::protocol::{BaseFrame, RequestHeader, ResponseHeader};

#[derive(Debug, Clone)]
pub struct HeartbeatContext {
    pub req_header: RequestHeader,
    pub resp_header: ResponseHeader,
    pub info: String,
}

impl BaseFrame for HeartbeatContext {
    fn request_header(&self) -> &RequestHeader { &self.req_header }
    fn request_header_mut(&mut self) -> &mut RequestHeader { &mut self.req_header }
    fn response_header(&self) -> &ResponseHeader { &self.resp_header }
    fn response_header_mut(&mut self) -> &mut ResponseHeader { &mut self.resp_header }

    fn serialize_request_body(&mut self) -> Vec<u8> {
        Vec::new()
    }

    fn deserialize_response_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let mut bs = BinaryStream::from_vec(data.to_vec());
        self.info = bs.get_string(10)?;
        Ok(())
    }
}

impl HeartbeatContext {
    pub fn new() -> Self {
        let mut header = RequestHeader::new(STD_HEARTBEAT, FLAG_UNCOMPRESSED);
        header.sequence_id = get_sequence_id();
        header.packet_ctrl = 0x02;
        Self {
            req_header: header,
            resp_header: ResponseHeader::new(),
            info: String::new(),
        }
    }
}
